### PR TITLE
Use peer recovery retention leases for indices without soft-deletes

### PR DESCRIPTION
**Pull request by dnhatn**
Today, the replica allocator uses peer recovery retention leases to select the best-matched copies when allocating replicas of indices with soft-deletes. We can employ this mechanism for indices without soft-deletes because the retaining sequence number of a PRRL is the persisted global checkpoint (plus one) of that copy. If the primary and replica have the same retaining sequence number, then we should be able to perform a noop recovery. The reason is that we must be retaining translog up to the local checkpoint of the safe commit, which is at most the global checkpoint of either copy). The only limitation is that we might not cancel ongoing file-based recoveries with PRRLs for noop recoveries. We can't make the translog retention policy comply with PRRLs. We also have this problem with soft-deletes if a PRRL is about to expire.

A nice [side-effect](https://github.com/elastic/elasticsearch/commit/bda80cde2422eae99934e3e9f86cd51c699fe623) of this is that we can turn off the translog retention once all shards started. However, I prefer leaving translog disconnect to PRRLs.

Relates #45136
Relates #46959

### CODE REVIEW

**Review by ywelsch — COMMENTED** [review]
This is looking good. I've left one question about strengthening the assertions in the tests.

File: test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java [hunk 1]
```diff
@@ -1139,9 +1139,10 @@ public void ensurePeerRecoveryRetentionLeasesRenewedAndSynced(String index) thro
                     assertNotNull(globalCheckpoint);
                     @SuppressWarnings("unchecked") List<Map<String, ?>> retentionLeases =
                         (List<Map<String, ?>>) XContentMapValues.extractValue("retention_leases.leases", copy);
-                    if (retentionLeases == null) {
+                    if (alwaysExists == false && retentionLeases == null) {
                         continue;
                     }
+                    assertNotNull(retentionLeases);
                     for (Map<String, ?> retentionLease : retentionLeases) {
                         if (((String) retentionLease.get("id")).startsWith("peer_recovery/")) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java]: ++. Adjusted in https://github.com/elastic/elasticsearch/pull/50351/commits/6071abdc20f0933f0bfd1c1a87608e2821f883c1.

**Comment by dnhatn** [comment]
run elasticsearch-ci/packaging-sample-matrix-unix

File: qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java [hunk 2]
```diff
@@ -1278,7 +1278,7 @@ public void testOperationBasedRecovery() throws Exception {
                 }
             }
             flush(index, true);
-            ensurePeerRecoveryRetentionLeasesRenewedAndSynced(index);
+            ensurePeerRecoveryRetentionLeasesRenewedAndSynced(index, false);
```
Comment (ywelsch) [inline qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java]: Should we set `alwaysExists ` to `true` if `minimumNodeVersion()` is on or after 7.6 (after backport) (here as well as in other places)?

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
@elasticmachine update branch

**Comment by dnhatn** [comment]
@elasticmachine update branch

File: qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java (hunk 2 above)
Comment (dnhatn) [reply qa/full-cluster-restart/src/test/java/org/elasticsearch/upgrades/FullClusterRestartIT.java]: Sure, will do.

**Comment by dnhatn** [comment]
Thanks Yannick!

**Pull request merged**